Prepare release 0.11.0.0 (#268)
haskell/bytestring: Changelog.md

[0.11.0.0] — September 2020

>  * [Change internal representation of `ByteString`, removing offset](https://github.com/haskell/bytestring/pull/175)

Reviewer: It might be good to add a sentence about the `PS` pattern synonym here. E.g.

```suggestion
 * [Change internal representation of `ByteString`, removing offset](https://github.com/haskell/bytestring/pull/175)
   * The old `PS` constructor has been turned into a pattern synonym that is available with GHC >= 8.0 for backwards compatibility.
```

 * [Remove deprecated functions `findSubstring` and `findSubstrings`](https://github.com/haskell/bytestring/pull/181)
 * [Speed up sorting of short strings](https://github.com/haskell/bytestring/pull/267)

>  * [Improve handling of literal strings](https://github.com/haskell/bytestring/pull/132)

Reviewer: ```suggestion
 * [Improve handling of literal strings in `Data.ByteString.Builder`](https://github.com/haskell/bytestring/pull/132)
```

 * [Add `unsafePackLiteral` to `Data.ByteString.Internal`](https://github.com/haskell/bytestring/pull/191)
   When possible, use
   [known-key](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/compiler/wired-in)
   variant of C's `strlen` from `GHC.CString` that supports constant folding.
   This results in better generated code when an ASCII-only string literal

>    is desugared to `ByteString` by the `OverloadedStrings` extension.

Reviewer: Maybe move this description into a (sub)list item or just add a dot after

    Add `unsafePackLiteral` to `Data.ByteString.Internal`

?

Author: I wonder whether this entry is too long and too specific for a changelog. My understanding is that the main benefit is not to provide a new function (not that we really expect end users to use it directly). What about 
"Improve opportunities for optimizations when a function scrutinizes the length of a `ByteString`"?

Reviewer: Yeah, this could be clearer and probably shorter.

How about

> Compute length of literal strings at compile time
> * This improves optimization opportunities for functions that scrutinize the length of a `ByteString`.

(CC @andrewthad)

   * The old `PS` constructor has been turned into a pattern synonym that is available with GHC >= 8.0 for backwards compatibility.
 * [Remove deprecated functions `findSubstring` and `findSubstrings`](https://github.com/haskell/bytestring/pull/181)
 * [Speed up sorting of short strings](https://github.com/haskell/bytestring/pull/267)
 * [Improve handling of literal strings in `Data.ByteString.Builder`](https://github.com/haskell/bytestring/pull/132)
 * [Compute length at compile time for literal strings](https://github.com/haskell/bytestring/pull/191)
   * This improves optimization opportunities for functions that scrutinize the length of a `ByteString`.
 * [Add `indexMaybe` and synonym `(!?)` for indexing that returns `Maybe`](https://github.com/haskell/bytestring/pull/261)
 * [Add rewrite rules for `{take,drop}While ({=,/}= x)`](https://github.com/haskell/bytestring/pull/275)
 * [Add rewrite rules for `any (== x)` and `all (/= x)`](https://github.com/haskell/bytestring/pull/273)
 * [Add rewrite rules for `findInd{ex,ices} (== x)`](https://github.com/haskell/bytestring/pull/270)

>  * [Improve folds to pass less argument on each recursive invocation](https://github.com/haskell/bytestring/pull/273)

Reviewer: ```suggestion
 * [Improve folds to pass fewer arguments on each recursive invocation](https://github.com/haskell/bytestring/pull/273)
```
